File: src/services/genres.ts
import { apiClient } from "./api-client";
import { Track } from "../types";

export interface TrackSuggestion {
  id: number;
  title: string;
  artist: string;
  bpm: number;
  filepath: string;
  current_genre?: string;
}

export interface GroupedSuggestion {
  parent_track: Track;
  suggestions: TrackSuggestion[];
}

export interface GroupedSuggestionSummary {
  parent_track: Track;
  suggestion_count: number;
}

export interface GenreBatchUpdateRequest {
  parent_track_id: number;
  target_track_ids: number[];
}

export type AnalysisMode = "genre" | "subgenre" | "both";

export interface GenreAnalysisResult {
  genre: string;
  subgenre?: string;
  reason: string;
  confidence: string;
}

export interface GenreUpdateResult {
  track_id: number;
  title: string;
  artist: string;
  old_genre: string;
  new_genre: string;
}

export interface GenreCleanupGroup {
  primary_genre: string;
  variant_genres: string[];
  track_count: number;
  suggestions: TrackSuggestion[];
}

export const genreService = {
  getAllGenres: async (): Promise<string[]> => {
    return apiClient.get<string[]>("/genres/list");
  },

  getAllSubgenres: async (): Promise<string[]> => {
    return apiClient.get<string[]>("/genres/subgenres");
  },

  getUnknownTracks: async (
    offset: number = 0,
    limit: number = 50,
    mode: AnalysisMode = "genre"
  ): Promise<Track[]> => {
    return apiClient.get<Track[]>("/genres/unknown", { offset, limit, mode });
  },

  getAllUnknownTrackIds: async (mode: AnalysisMode = "genre"): Promise<number[]> => {
    return apiClient.get<number[]>("/genres/unknown-ids", { mode });
  },

  getGroupedSuggestions: async (
    offset: number = 0,
    limit: number = 10,
    threshold: number = 0.85,
    mode: AnalysisMode = "genre"
  ): Promise<GroupedSuggestionSummary[]> => {
    return apiClient.get<GroupedSuggestionSummary[]>(
      "/genres/grouped-suggestions",
      {
        offset,
        limit,
        threshold,
        mode,
      }
    );
  },

  getSuggestionsForTrack: async (
    trackId: number,
    threshold: number = 0.85
  ): Promise<TrackSuggestion[]> => {
    return apiClient.get<TrackSuggestion[]>(
      `/genres/grouped-suggestions/${trackId}`,
      {
        threshold,
      }
    );
  },

  batchUpdateGenres: async (
    parentTrackId: number,
    targetTrackIds: number[]
  ): Promise<{ updated_count: number; genre: string }> => {
    return apiClient.post<{ updated_count: number; genre: string }>(
      "/genres/batch-update",
      {
        parent_track_id: parentTrackId,
        target_track_ids: targetTrackIds,
      }
    );
  },

  getCleanupSuggestions: async (mode: AnalysisMode = "genre"): Promise<GenreCleanupGroup[]> => {
    return apiClient.get<GenreCleanupGroup[]>("/genres/cleanup-suggestions", { mode });
  },

  executeCleanup: async (
    targetGenre: string,
    trackIds: number[],
    mode: AnalysisMode = "genre"
  ): Promise<{ updated_count: number; genre: string }> => {
    return apiClient.post<{ updated_count: number; genre: string }>(
      "/genres/cleanup-execute",
      {
        target_genre: targetGenre,
        track_ids: trackIds,
        mode,
      }
    );
  },

  applyGenresToFiles: async (
    trackIds: number[]
  ): Promise<{ success: number; failed: number }> => {
    return apiClient.post<{ success: number; failed: number }>(
      "/genres/apply-to-files",
      { track_ids: trackIds }
    );
  },
};

File: backend/api/schemas/genres.py
from pydantic import BaseModel
from typing import List, Optional
from enum import Enum
from .track import TrackRead

class AnalysisMode(str, Enum):
    GENRE = "genre"
    SUBGENRE = "subgenre"
    BOTH = "both"

class GenreBatchUpdateRequest(BaseModel):
    parent_track_id: int
    target_track_ids: List[int]

class GenreBatchUpdateResponse(BaseModel):
    updated_count: int
    genre: str

class GenreCleanupRequest(BaseModel):
    target_genre: str
    track_ids: List[int]
    mode: AnalysisMode = AnalysisMode.GENRE

class GenreLLMAnalyzeRequest(BaseModel):
    track_id: int
    overwrite: bool = False
    mode: AnalysisMode = AnalysisMode.BOTH

class GenreBatchLLMAnalyzeRequest(BaseModel):
    track_ids: List[int]
    mode: AnalysisMode = AnalysisMode.BOTH
    overwrite: bool = False

class GenreUpdateResult(BaseModel):
    track_id: int
    title: str
    artist: str
    old_genre: str
    new_genre: str

class GenreAnalysisResponse(BaseModel):
    genre: str
    subgenre: str = ""
    reason: str
    confidence: str

class GenreBatchAnalysisItem(GenreAnalysisResponse):
    track_id: int

class GenreBatchAnalysisResponse(BaseModel):
    results: List[GenreBatchAnalysisItem]

class TrackSuggestion(BaseModel):
    id: int
    title: str
    artist: str
    bpm: float
    filepath: str
    current_genre: Optional[str] = None

class GroupedSuggestion(BaseModel):
    parent_track: TrackRead
    suggestions: List[TrackSuggestion]

class GroupedSuggestionSummary(BaseModel):
    parent_track: TrackRead
    suggestion_count: int

class GenreCleanupGroup(BaseModel):
    primary_genre: str
    variant_genres: List[str]
    track_count: int
    suggestions: List[TrackSuggestion]

class GenreApplyRequest(BaseModel):
    track_ids: List[int]